Tests/JWECompressionTests.swift

```diff
@@ -0,0 +1,58 @@
+//
+//  JWECompressionTests.swift
+//  Tests
+//
+//  Created by Florian Häser on 24.12.18.
+//
+
+import XCTest
+@testable import JOSESwift
+
+class JWECompressionTests: RSACryptoTestCase {
+    
+    let data = "So Secret! 🔥🌵".data(using: .utf8)!
+    
+    let jweSerializedNotSupportedZipHeaderValue = "eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2Q0JDLUhTNTEyIiwiemlwIjoiR1pJUCJ9..I62WKaitlHoJ-Kz3zvQ-Tw.Tu9hk_AMRMRTc8ggsoWifFS979Nz8xt-xx4FpF6waeE.w7c8eAGXpUD3tNskLzdl17s4vsCCSUwe5bRFpJg1kUs"
+    
+    @available(*, deprecated)
+    func testRoundtripWithLegacyDecrypter() {
+        let symmetricKey = try! SecureRandom.generate(count: SymmetricKeyAlgorithm.A256CBCHS512.keyLength)
+        
+        var header = JWEHeader(algorithm: .direct, encryptionAlgorithm: .A256CBCHS512)
+        header.zip = "DEF"
+        
+        let payload = Payload(data)
+        let encrypter = Encrypter(keyEncryptionAlgorithm: .direct, encryptionKey: symmetricKey, contentEncyptionAlgorithm: .A256CBCHS512)!
+        
+        let jwe = try! JWE(header: header, payload: payload, encrypter: encrypter)
+        let serialization = jwe.compactSerializedString
+        
+        try! XCTAssertEqual(JWE(compactSerialization: serialization).decrypt(with: symmetricKey).data(), data)
+    }
+    
+    func testRoundtrip() {
+        let symmetricKey = try! SecureRandom.generate(count: SymmetricKeyAlgorithm.A256CBCHS512.keyLength)
+        
+        var header = JWEHeader(algorithm: .direct, encryptionAlgorithm: .A256CBCHS512)
+        header.zip = "DEF"
+        
+        let payload = Payload(data)
+        let encrypter = Encrypter(keyEncryptionAlgorithm: .direct, encryptionKey: symmetricKey, contentEncyptionAlgorithm: .A256CBCHS512)!
+        
+        let jwe = try! JWE(header: header, payload: payload, encrypter: encrypter)
+        let serialization = jwe.compactSerializedString
+        
+        let decrypter = Decrypter(keyDecryptionAlgorithm: .direct, decryptionKey: symmetricKey, contentDecryptionAlgorithm: .A256CBCHS512)!
+        
+        try! XCTAssertEqual(JWE(compactSerialization: serialization).decrypt(using: decrypter).data(), data)
+    }
+
+    // Note this test only works as long as the guard statement is before the actual decryption operation
+    func testDecryptWithNotSupportedZipHeaderValue() {
```

Comment:
This test throws (and therefore succeeds) because you generate a brand new random key to decrypt with, not because of an unknown compression algorithm. 😄 ⚠️ 

With the modifications above (error propagation in the compression factory and only returning the none-compressor if explicitly requested), this should work. Please make sure to check exactly which error is thrown. For that you'll need to [make the errors equatable in the tests](https://github.com/airsidemobile/JOSESwift/blob/f40e59e7d3538b81f91571cd5cc4153903868f9a/Tests/JWSValidationTests.swift#L241) and then use something like that:

``` swift
XCTAssertThrowsError(/* ... */ ) { error in
     XCTAssertEqual(error as! JOSESwiftError, JOSESwiftError.decryptionFailed(description: /* ... */))
}
```


Let me know if I can help you with that.

Reply:
Yeah you are right @daniel-mohemian. After applying all of your suggestions I ended up in using that assert.
Question here is it OK to assert for
```swift
XCTAssertEqual(error as! JOSESwiftError,
              JOSESwiftError.encryptingFailed(description:
                   "The operation couldn’t be completed.
                    (JOSESwift.JOSESwiftError error 22.)"
                ))
```
At first sight the number seemed meaningless until I realized that is is the order of the error in the `JOSESwiftError` enumeration.